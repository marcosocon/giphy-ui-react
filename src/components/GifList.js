import React from "react";
import { uniqueId, isEmpty, find } from 'lodash';

import { Container, Grid, Image, Button, Header } from 'semantic-ui-react';

function GifList({ title, favorites, gifsToShow, onAddFavorite, onRemoveFavorite }) {
    const showFavoriteHandler = (item) => {
        const isAlreadyAFavorite = !!find(favorites, (el) => el.id === item.id);
        return (
            <Button
                circular
                icon="heart"
                className="like"
                color={isAlreadyAFavorite ? 'black' : 'red'}
                onClick={() => isAlreadyAFavorite ? onRemoveFavorite(item) : onAddFavorite(item)}
            />
        )
    }

    return gifsToShow && !isEmpty(gifsToShow) ? (
        <Container>
            <Header
                as="h1"
                className="h1">
                {title}
            </Header>
            <Grid columns={4}>
                {gifsToShow.map((item, key) => {
                    return (
                        <Grid.Column key={uniqueId("trending-")}>
                            <Image
                                src={item.images.fixed_height.url}
                                as="a"
                                href={item.images.original.url}
                                target="_blank"
                                bordered
                            />
                            {showFavoriteHandler(item)}
                        </Grid.Column>
                    )
                })}
            </Grid>
        </Container>
    ) : null;
};

export default GifList;