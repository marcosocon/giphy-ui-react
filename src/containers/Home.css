.h1.ui.header {
  font-size: 1.5rem;
}

.headerWrapper {
  padding: 3rem 0 !important;
}

.ui .search > .ui.input {
  width: 100%;
}

.like {
  position: absolute;
  right: 7%;
  top: 15%;
}

.searchLoader {
  height: 75vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
}