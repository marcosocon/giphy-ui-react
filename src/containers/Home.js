import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Container, Grid, Header, Input } from 'semantic-ui-react';
import { debounce, isEmpty } from 'lodash';

import { fetchGifsByKeyword } from '../actions/search';
import { addFavorite, removeFavorite } from '../actions/favorites';
import { fetchTrendingGifs } from '../actions/trending';

import './Home.css';
import GifList from '../components/GifList';
import Loader from '../components/Loader';

export class HomeContainer extends Component {
    constructor(props) {
        super(props);
        this.state = {
            searchStr: ''
        };
    }

    componentDidMount() {
        this.props.fetchTrendingGifs();
    }

    debounceSearch = debounce(() => this.props.fetchGifsByKeyword(this.state.searchStr), 500);

    handleChange = (e) => {
        this.setState({ [e.target.name]: e.target.value });
        this.debounceSearch();
    }

    render() {
        const { trending, search, results, addFavorite, removeFavorite, favorites } = this.props;
        const gifs = search.status === 0 || isEmpty(this.state.searchStr) ? trending : results;
        const title = search.status === 0 || isEmpty(this.state.searchStr) ? 'Trending' : 'Results'
        return (
            <Container>
                <GifList
                    title='Favorites'
                    favorites={favorites}
                    gifsToShow={favorites}
                    onAddFavorite={addFavorite}
                    onRemoveFavorite={removeFavorite} />
                <Grid>
                    <Grid.Row className="headerWrapper">
                        <Grid.Column
                            width={3}
                            verticalAlign="middle">
                            <Header
                                as="h1"
                                className="h1">
                                GIPHY UI
                            </Header>
                        </Grid.Column>
                        <Grid.Column width={13}>
                            <Input
                                icon='search'
                                placeholder='Write a keyword to search a GIF'
                                name='searchStr'
                                onChange = {this.handleChange}
                                loading={search.loading}
                                results={results}
                                value={this.state.searchStr}
                            />
                        </Grid.Column>
                    </Grid.Row>
                </Grid>
                {search.loading ? (
                    <Loader classes="searchLoader" />
                ) : (
                    <GifList
                        title={title}
                        favorites={favorites}
                        gifsToShow={gifs}
                        onAddFavorite={addFavorite}
                        onRemoveFavorite={removeFavorite}
                    />
                )}
            </Container>
        );
    }
};

export const mapStateToProps = (state) => ({
    favorites: state.gifsReducer.favoriteGifs,
    trending: state.gifsReducer.trendingGifs,
    results: state.gifsReducer.searchResultGifs,
    search: state.searchReducer,
});

export const mapDispatchToProps = (dispatch) => ({
    fetchTrendingGifs: () => dispatch(fetchTrendingGifs()),
    fetchGifsByKeyword: (payload) => dispatch(fetchGifsByKeyword(payload)),
    removeFavorite: (payload) => dispatch(removeFavorite(payload)),
    addFavorite: (payload) => dispatch(addFavorite(payload)),
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(HomeContainer);