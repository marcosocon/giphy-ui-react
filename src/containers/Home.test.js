import React from 'react';
import { shallow, mount } from 'enzyme';
import { HomeContainer } from './Home';
import GifList from '../components/GifList';

describe('<Home />', () => {
  const props = {
    search: {
      loading: true,
      status: 0,
    },
    fetchTrendingGifs: jest.fn(),
  };

  it('render itself', () => {
    const wrapper = shallow(<HomeContainer {...props} />)
    expect(wrapper).toMatchSnapshot();
  });

  it('should show 2 GifList when loading is false', () => {
    const props = {
      search: {
        loading: false,
        status: 0,
      },
      fetchTrendingGifs: jest.fn(),
    }
    const wrapper = shallow(<HomeContainer {...props} />);
    expect(wrapper.find(GifList)).toHaveLength(2);
  })
});
