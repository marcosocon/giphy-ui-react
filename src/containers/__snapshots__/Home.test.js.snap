// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> render itself 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HomeContainer
    fetchTrendingGifs={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    search={
      Object {
        "loading": true,
        "status": 0,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <GifList
          title="Favorites"
        />,
        <Grid>
          <GridRow
            className="headerWrapper"
          >
            <GridColumn
              verticalAlign="middle"
              width={3}
            >
              <Header
                as="h1"
                className="h1"
              >
                GIPHY UI
              </Header>
            </GridColumn>
            <GridColumn
              width={13}
            >
              <Input
                icon="search"
                loading={true}
                name="searchStr"
                onChange={[Function]}
                placeholder="Write a keyword to search a GIF"
                type="text"
                value=""
              />
            </GridColumn>
          </GridRow>
        </Grid>,
        <Loader
          classes="searchLoader"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "favorites": undefined,
          "gifsToShow": undefined,
          "onAddFavorite": undefined,
          "onRemoveFavorite": undefined,
          "title": "Favorites",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <GridRow
            className="headerWrapper"
          >
            <GridColumn
              verticalAlign="middle"
              width={3}
            >
              <Header
                as="h1"
                className="h1"
              >
                GIPHY UI
              </Header>
            </GridColumn>
            <GridColumn
              width={13}
            >
              <Input
                icon="search"
                loading={true}
                name="searchStr"
                onChange={[Function]}
                placeholder="Write a keyword to search a GIF"
                type="text"
                value=""
              />
            </GridColumn>
          </GridRow>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <GridColumn
                verticalAlign="middle"
                width={3}
              >
                <Header
                  as="h1"
                  className="h1"
                >
                  GIPHY UI
                </Header>
              </GridColumn>,
              <GridColumn
                width={13}
              >
                <Input
                  icon="search"
                  loading={true}
                  name="searchStr"
                  onChange={[Function]}
                  placeholder="Write a keyword to search a GIF"
                  type="text"
                  value=""
                />
              </GridColumn>,
            ],
            "className": "headerWrapper",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Header
                  as="h1"
                  className="h1"
                >
                  GIPHY UI
                </Header>,
                "verticalAlign": "middle",
                "width": 3,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "as": "h1",
                  "children": "GIPHY UI",
                  "className": "h1",
                },
                "ref": null,
                "rendered": "GIPHY UI",
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Input
                  icon="search"
                  loading={true}
                  name="searchStr"
                  onChange={[Function]}
                  placeholder="Write a keyword to search a GIF"
                  type="text"
                  value=""
                />,
                "width": 13,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "icon": "search",
                  "loading": true,
                  "name": "searchStr",
                  "onChange": [Function],
                  "placeholder": "Write a keyword to search a GIF",
                  "results": undefined,
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "classes": "searchLoader",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <GifList
            title="Favorites"
          />,
          <Grid>
            <GridRow
              className="headerWrapper"
            >
              <GridColumn
                verticalAlign="middle"
                width={3}
              >
                <Header
                  as="h1"
                  className="h1"
                >
                  GIPHY UI
                </Header>
              </GridColumn>
              <GridColumn
                width={13}
              >
                <Input
                  icon="search"
                  loading={true}
                  name="searchStr"
                  onChange={[Function]}
                  placeholder="Write a keyword to search a GIF"
                  type="text"
                  value=""
                />
              </GridColumn>
            </GridRow>
          </Grid>,
          <Loader
            classes="searchLoader"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "favorites": undefined,
            "gifsToShow": undefined,
            "onAddFavorite": undefined,
            "onRemoveFavorite": undefined,
            "title": "Favorites",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <GridRow
              className="headerWrapper"
            >
              <GridColumn
                verticalAlign="middle"
                width={3}
              >
                <Header
                  as="h1"
                  className="h1"
                >
                  GIPHY UI
                </Header>
              </GridColumn>
              <GridColumn
                width={13}
              >
                <Input
                  icon="search"
                  loading={true}
                  name="searchStr"
                  onChange={[Function]}
                  placeholder="Write a keyword to search a GIF"
                  type="text"
                  value=""
                />
              </GridColumn>
            </GridRow>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <GridColumn
                  verticalAlign="middle"
                  width={3}
                >
                  <Header
                    as="h1"
                    className="h1"
                  >
                    GIPHY UI
                  </Header>
                </GridColumn>,
                <GridColumn
                  width={13}
                >
                  <Input
                    icon="search"
                    loading={true}
                    name="searchStr"
                    onChange={[Function]}
                    placeholder="Write a keyword to search a GIF"
                    type="text"
                    value=""
                  />
                </GridColumn>,
              ],
              "className": "headerWrapper",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <Header
                    as="h1"
                    className="h1"
                  >
                    GIPHY UI
                  </Header>,
                  "verticalAlign": "middle",
                  "width": 3,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "as": "h1",
                    "children": "GIPHY UI",
                    "className": "h1",
                  },
                  "ref": null,
                  "rendered": "GIPHY UI",
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <Input
                    icon="search"
                    loading={true}
                    name="searchStr"
                    onChange={[Function]}
                    placeholder="Write a keyword to search a GIF"
                    type="text"
                    value=""
                  />,
                  "width": 13,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "icon": "search",
                    "loading": true,
                    "name": "searchStr",
                    "onChange": [Function],
                    "placeholder": "Write a keyword to search a GIF",
                    "results": undefined,
                    "type": "text",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "classes": "searchLoader",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
